### PR TITLE
add benchmark failure check

### PR DESCRIPTION
**Pull request by xspronken**
Add a job that fails the workflow if a benchmark fails.

### CODE REVIEW

**Comment by hodgestar** [comment]
There is a much simpler and better way to achieve this because `pytest.main` returns the exit code.

```python
def run_benchmarks(args):
    "Run pytest benchmark with sensible defaults."
    pytest.main(
```
In `run_benchmark` above, change the first line to `return pytest.main(` so the exit code is returned.

Then in your own `def main` function change
```python
    run_benchmarks(other_args)
    benchmark_latest = get_latest_benchmark()
    add_packages_to_json(benchmark_latest)
```
Save the exit code in a variable, don't save the benchmarks if the exit code is not 0, and return the exit code from your function.

Finally, change the end of your script to:
```python
if __name__ == "__main__":
    sys.exit(main())
```

**Review by hodgestar — CHANGES_REQUESTED** [review]
There is a much simpler solution that just makes sure the correct exit code is returned. That will make the script do the right thing everywhere, and not require dubious matching of stdout.

I wrote a fuller description of the better solution as a comment.

**Comment by xspronken** [comment]
This is indeed much better. I don't think there is a need to only save the benchmarks on exit code 0. They are saved automatically when running:
```
pytest.main(
        [
            "qutip_benchmark/benchmarks",
            "--benchmark-only",
            "--benchmark-columns=Mean,StdDev,rounds,Iterations",
            "--benchmark-sort=name",
            "--benchmark-autosave",
            "-Wdefault",
        ]
        + args
    )
```
This means they will be saved regardless of the exit code. I could delete the file if exit code is not 0 but i don't think this is useful, since the entire workflow stops on failure and all the subsequent jobs are cancelled. So regardless of whether the benchmarks are saved or not, if they fail they will not be uploaded to S3, nor will an artefact be created (see [here](https://github.com/qutip/qutip-benchmark/runs/8292381873?check_suite_focus=true)).

**Comment by hodgestar** [comment]
If they are written already, happy for them to be kept.

**Review by Ericgig — APPROVED** [review]
Look good, but you will probably have a conflict merging this and #29.

File: .github/workflows/nightly_benchmarks.yml [hunk 1]
```diff
@@ -72,9 +71,9 @@ jobs:
           AWS_SECRET_ACCESS_KEY: ${{ secrets.QUTIP_BENCHMARK_S3_SECRET }}
           AWS_EC2_METADATA_DISABLED: true
 
-      - name: run benchmarks
+      - name: Run benchmarks
         run: |
-          python -m qutip_benchmark.cli.run_benchmarks -m nightly -v 
+          python benchmarks/benchmarks.py -v -m nightly
```
Comment (hodgestar) [inline .github/workflows/nightly_benchmarks.yml:76]:
```suggestion
          -m qutip_benchmark.cli.run_benchmarks -m nightly -v
```
Perhaps a bad merge?

File: .github/workflows/nightly_benchmarks.yml [hunk 2]
```diff
@@ -72,9 +71,9 @@ jobs:
           AWS_SECRET_ACCESS_KEY: ${{ secrets.QUTIP_BENCHMARK_S3_SECRET }}
           AWS_EC2_METADATA_DISABLED: true
 
-      - name: run benchmarks
+      - name: Run benchmarks
         run: |
-          python -m qutip_benchmark.cli.run_benchmarks -m nightly -v 
+          -m qutip_benchmark.cli.run_benchmarks -m nightly -v
```
Comment (hodgestar) [inline .github/workflows/nightly_benchmarks.yml:76]:
```suggestion
          python -m qutip_benchmark.cli.run_benchmarks -m nightly -v
```

**Review by hodgestar — APPROVED** [review]

**Pull request merged**